lib/src/controls/flyouts/flyout.dart

```diff
@@ -0,0 +1,779 @@
+    // we perform this check before all the calculation to save computing time
+    final pas = configuration.preferredMode
+        ._getAvailableSpace(targetOffset, rootSize, margin)
+        .biggest;
+    if (pas.height >= availableSpace) {
+      return configuration.preferredMode;
+    }
+
+    final isLeftPreferred = [
+      FlyoutPlacementMode.left,
+      FlyoutPlacementMode.topLeft,
+      FlyoutPlacementMode.bottomLeft,
+    ].contains(configuration.preferredMode);
+    final isCenterPreferred = [
+      FlyoutPlacementMode.topCenter,
+      FlyoutPlacementMode.bottomCenter
+    ].contains(configuration.preferredMode);
+    final isRightPreferred = [
+      FlyoutPlacementMode.right,
+      FlyoutPlacementMode.topRight,
+      FlyoutPlacementMode.bottomRight
+    ].contains(configuration.preferredMode);
+
+    final bas = FlyoutPlacementMode.bottomCenter
+        ._getAvailableSpace(
+          targetOffset,
+          rootSize,
+          margin,
+        )
+        .biggest;
+
+    if (bas.height >= availableSpace) {
+      if (isLeftPreferred) return FlyoutPlacementMode.bottomLeft;
+      if (isCenterPreferred) return FlyoutPlacementMode.bottomCenter;
+      if (isRightPreferred) return FlyoutPlacementMode.bottomRight;
+    }
+
+    final tas = FlyoutPlacementMode.topCenter
+        ._getAvailableSpace(
+          targetOffset,
+          rootSize,
+          margin,
+        )
+        .biggest;
+
+    if (tas.height >= availableSpace) {
+      if (isLeftPreferred) return FlyoutPlacementMode.topLeft;
+      if (isCenterPreferred) return FlyoutPlacementMode.topCenter;
+      if (isRightPreferred) return FlyoutPlacementMode.topRight;
+    }
+
+    return configuration.preferredMode;
+  }
+}
+
+extension on double {
+  /// If negative, 0 is returned
+  double _ensurePositive() {
+    if (isNegative) return 0;
+    return this;
+  }
+}
+
+class FlyoutAutoConfiguration {
+  /// The amount of necessary available space.
+  ///
+  /// If not provided, it falls back to the flyout size
+  final double? autoAvailableSpace;
+
+  /// Whether the flyout should be displayed horizontally
+  ///
+  /// If true, [preferredMode] must be either .left or .right
+  final bool horizontal;
+
+  /// The preferred mode
+  final FlyoutPlacementMode preferredMode;
+
+  /// The configuration for flyout auto mode
+  FlyoutAutoConfiguration({
+    this.autoAvailableSpace,
+    bool? horizontal,
+    required this.preferredMode,
+  })  : assert(preferredMode != FlyoutPlacementMode.auto),
+        assert(
+          horizontal != null && horizontal
+              ? preferredMode == FlyoutPlacementMode.left ||
+                  preferredMode == FlyoutPlacementMode.right
+              : true,
+          'If the mode horizontal, preferredMode must either be left or right',
+        ),
+        assert(autoAvailableSpace == null || !autoAvailableSpace.isNegative),
+        horizontal = horizontal ??
+            [FlyoutPlacementMode.left, FlyoutPlacementMode.right]
```

Comment:
Prefer having this as a getter so you can guarantee const-ness of this constructor